### ISSUE TITLE
Package does not compile on Raspbian Stretch (Raspberry Pi 3)

### ISSUE THREAD

**Issue by darioteixeira** [mirage/ocaml-uri#106]
When running `opam install uri` on a Raspberry Pi 3 (running Raspbian Stretch), I get the following error:

```
#=== ERROR while compiling uri.1.9.4 ==========================================#
# command      jbuilder build -p uri -j 3
# path         /home/pi/.opam/4.04.1/.opam-switch/build/uri.1.9.4
# exit-code    1
# env-file     /home/pi/.opam/log/uri-1642-e7966f.env
# output-file  /home/pi/.opam/log/uri-1642-e7966f.out
### output ###
# [...]
# /tmp/camlasmfc3e02.s:188838: Error: offset out of range
# /tmp/camlasmfc3e02.s:188841: Error: offset out of range
# /tmp/camlasmfc3e02.s:188845: Error: offset out of range
# /tmp/camlasmfc3e02.s:188860: Error: offset out of range
# /tmp/camlasmfc3e02.s:188864: Error: offset out of range
# /tmp/camlasmfc3e02.s:188871: Error: offset out of range
# /tmp/camlasmfc3e02.s:188892: Error: offset out of range
# /tmp/camlasmfc3e02.s:188896: Error: offset out of range
# /tmp/camlasmfc3e02.s:188900: Error: offset out of range
# /tmp/camlasmfc3e02.s:188917: Error: offset out of range
# File "etc/uri_services_full.ml", line 1:
# Error: Assembler error, input left in file /tmp/camlasmfc3e02.s
```

Note that compilation on an AMD64 machine completes without error. Is there perhaps something funky with my particular Raspbian installation, or does anyone else encounter the same problem?
(I can post the assembly file on `/tmp` if that helps.)

**Comment by dbuenzli** [comment]
This is unlikely but who knows could it be that you run out of disk space and the `.s` gets truncated ?

**Comment by darioteixeira** [comment]
> This is unlikely but who knows could it be that you run out of disk space and the .s gets truncated ?

Good guess, but that doesn't seem to be the case. Anyway, I forgot to mention that the compilation error occurs with version 1.9.4. Version 1.9.2 compiles fine, however.

**Comment by alrev** [comment]
The same error with 4.05.0 compiler:
```
#=== ERROR while installing uri.1.9.4 =========================================#
# opam-version 1.2.2
# os           linux
# command      jbuilder build -p uri -j 1
# path         /home/pi/.opam/4.05.0/build/uri.1.9.4
# compiler     4.05.0
# exit-code    1
# env-file     /home/pi/.opam/4.05.0/build/uri.1.9.4/uri-10038-21b540.env
# stdout-file  /home/pi/.opam/4.05.0/build/uri.1.9.4/uri-10038-21b540.out
# stderr-file  /home/pi/.opam/4.05.0/build/uri.1.9.4/uri-10038-21b540.err
### stderr ###
# [...]
# /tmp/camlasm687d62.s:148303: Error: offset out of range
# /tmp/camlasm687d62.s:148320: Error: offset out of range
# /tmp/camlasm687d62.s:188141: Error: offset out of range
# /tmp/camlasm687d62.s:188148: Error: offset out of range
# /tmp/camlasm687d62.s:188169: Error: offset out of range
# /tmp/camlasm687d62.s:188173: Error: offset out of range
# /tmp/camlasm687d62.s:188177: Error: offset out of range
# /tmp/camlasm687d62.s:188194: Error: offset out of range
# File "etc/uri_services_full.ml", line 1:
# Error: Assembler error, input left in file /tmp/camlasm687d62.s



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions failed
  ∗  install uri 1.9.4
The following changes have been performed
  ∗  install base                    v0.9.3    
  ∗  install base-bytes              base      
  ∗  install base-num                base      
  ∗  install conf-m4                 1         
  ∗  install jbuilder                1.0+beta11
  ∗  install num                     0         
  ∗  install ocaml-compiler-libs     v0.9.0    
  ∗  install ocaml-migrate-parsetree 1.0.1     
  ∗  install ocamlbuild              0.11.0    
  ∗  install ocamlfind               1.7.3     
  ∗  install ppx_ast                 v0.9.1    
  ∗  install ppx_core                v0.9.0    
  ∗  install ppx_driver              v0.9.1    
  ∗  install ppx_metaquot            v0.9.0    
  ∗  install ppx_optcomp             v0.9.0    
  ∗  install ppx_sexp_conv           v0.9.0    
  ∗  install ppx_traverse_builtins   v0.9.0    
  ∗  install ppx_type_conv           v0.9.0    
  ∗  install re                      1.7.1     
  ∗  install result                  1.2       
  ∗  install sexplib                 v0.9.1    
  ∗  install stdio                   v0.9.0    
  ∗  install stringext               1.4.3     
```

**Comment by hannesm** [comment]
maybe related to https://github.com/ocaml/ocaml/pull/1250 ? i.e. does an `opam switch 4.06.0+pr1250` result in an environment where uri builds?

**Comment by alrev** [comment]
It seems ppx_ast.v0.9.1 in not compatible with 4.06.0 (4.06.0+pr1250 was used) currently:
```
#=== ERROR while installing ppx_ast.v0.9.1 ====================================#
# opam-version 1.2.2
# os           linux
# command      jbuilder build --only-packages ppx_ast --root . -j 1 @install
# path         /home/pi/.opam/4.06.0+pr1250/build/ppx_ast.v0.9.1
# compiler     4.06.0+pr1250
# exit-code    1
# env-file     /home/pi/.opam/4.06.0+pr1250/build/ppx_ast.v0.9.1/ppx_ast-25737-f19885.env
# stdout-file  /home/pi/.opam/4.06.0+pr1250/build/ppx_ast.v0.9.1/ppx_ast-25737-f19885.out
# stderr-file  /home/pi/.opam/4.06.0+pr1250/build/ppx_ast.v0.9.1/ppx_ast-25737-f19885.err
### stderr ###
# [...]
#     ocamlopt src/ppx_ast__Warn.{cmx,o}
#     ocamlopt src/ppx_ast__Ast.{cmx,o}
#       ocamlc src/ppx_ast__Ast_helper.{cmo,cmt}
#       ocamlc src/ppx_ast__Pprintast.{cmo,cmt}
#       ocamlc src/ppx_ast__Parser.{cmo,cmt}
#       ocamlc src/ppx_ast__Lexer.{cmi,cmo,cmt} (exit 2)
# (cd _build/default && /home/pi/.opam/4.06.0+pr1250/bin/ocamlc.opt -w -40 -open Ocaml_shadow -w -9-27-32 -g -bin-annot -I /home/pi/.opam/4.06.0+pr1250/lib/ocaml-compiler-libs/common -I /home/pi/.opam/4.06.0+pr1250/lib/ocaml-compiler-libs/shadow -I /home/pi/.opam/4.06.0+pr1250/lib/ocaml-migrate-parsetree -I /home/pi/.opam/4.06.0+pr1250/lib/ocaml/compiler-libs -no-alias-deps -I src -open Ppx_ast__ -o src/ppx_ast__Lexer.cmo -c -impl src/lexer.ml)
# File "src/lexer.mll", line 226, characters 4-64:
# Error: The constructor Warnings.Deprecated expects 3 argument(s),
#        but is applied here to 1 argument(s)



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install ppx_core      v0.9.0
  ∗  install ppx_driver    v0.9.1
  ∗  install ppx_metaquot  v0.9.0
  ∗  install ppx_optcomp   v0.9.0
  ∗  install ppx_sexp_conv v0.9.0
  ∗  install ppx_type_conv v0.9.0
  ∗  install uri           1.9.4 
The following actions failed
  ∗  install ppx_ast v0.9.1
The following changes have been performed
  ∗  install base                    v0.9.3    
  ∗  install base-bytes              base      
  ∗  install base-num                base      
  ∗  install conf-m4                 1         
  ∗  install jbuilder                1.0+beta11
  ∗  install num                     0         
  ∗  install ocaml-compiler-libs     v0.9.0    
  ∗  install ocaml-migrate-parsetree 1.0.1     
  ∗  install ocamlbuild              0.11.0    
  ∗  install ocamlfind               1.7.3     
  ∗  install ppx_traverse_builtins   v0.9.0    
  ∗  install re                      1.7.1     
  ∗  install result                  1.2       
  ∗  install sexplib                 v0.9.1    
  ∗  install stdio                   v0.9.0    
  ∗  install stringext               1.4.3     

```
I tried 4.06.0+trunk on x64 Ubuntu desktop, the similar error occured with ocaml-migrate-parsetree.1.0.1.

**Comment by orbifx** [comment]
I have the same error with another ARM chip. I checked `/tmp/`, which still contains some compilation artifacts, and it's not full.

A key part seems to be this:
```
/tmp/camlasm480e0c.s: Assembler messages:
/tmp/camlasm480e0c.s:52: Error: value of 68249 too large for field of 2 bytes at 30
/tmp/camlasm480e0c.s:73426: Error: value of 68255 too large for field of 2 bytes at 204542
```

Log file cropped to relevant section:


```
File "lib/uri.ml", line 306, characters 33-50:
Warning 52: Code should not depend on the actual values of
this constructor's arguments. They are only for information
and may change in future versions. (See manual section 8.5)
File "lib/uri.ml", line 321, characters 27-44:
Warning 52: Code should not depend on the actual values of
this constructor's arguments. They are only for information
and may change in future versions. (See manual section 8.5)
File "lib/uri.ml", line 568, characters 17-33:
Warning 3: deprecated: String.lowercase
Use String.lowercase_ascii instead.
    ocamlopt lib/uri_top.{cmx,o}
File "_none_", line 1:
Warning 58: no cmx file was found in path for module Toploop, and its interface was not compiled with -opaque
    ocamlopt etc/uri_services.{cmx,o}
    ocamlopt lib/uri.{a,cmxa}
    ocamlopt lib/uri_top.{a,cmxa}
    ocamlopt etc/uri_services.{a,cmxa}
    ocamlopt lib/uri.cmxs
    ocamlopt lib/uri_top.cmxs
    ocamlopt etc/uri_services.cmxs
      ocamlc etc/uri_services_full.{cmo,cmt}
      ocamlc etc/uri_services_full.cma
    ocamlopt etc/uri_services_full.{cmx,o} (exit 2)
(cd _build/default && /usr/bin/ocamlopt.opt -w -40 -g -I /home/fox/.opam/system/lib/bytes -I /home/fox/.opam/system/lib/re -I /home/fox/.opam/system/lib/sexplib -I /home/fox/.opam/system/lib/sexplib/0 -I /home/fox/.opam/system/lib/stringext -I /usr/lib/ocaml -I lib -no-alias-deps -I etc -o etc/uri_services_full.cmx -c -impl etc/uri_services_full.ml)
/tmp/camlasm480e0c.s: Assembler messages:
/tmp/camlasm480e0c.s:52: Error: value of 68249 too large for field of 2 bytes at 30
/tmp/camlasm480e0c.s:73426: Error: value of 68255 too large for field of 2 bytes at 204542
/tmp/camlasm480e0c.s:148307: Error: offset out of range
/tmp/camlasm480e0c.s:188181: Error: offset out of range
/tmp/camlasm480e0c.s:148267: Error: offset out of range
/tmp/camlasm480e0c.s:148274: Error: offset out of range
/tmp/camlasm480e0c.s:148295: Error: offset out of range
/tmp/camlasm480e0c.s:148299: Error: offset out of range
/tmp/camlasm480e0c.s:148303: Error: offset out of range
/tmp/camlasm480e0c.s:148320: Error: offset out of range
/tmp/camlasm480e0c.s:188141: Error: offset out of range
/tmp/camlasm480e0c.s:188148: Error: offset out of range
/tmp/camlasm480e0c.s:188169: Error: offset out of range
/tmp/camlasm480e0c.s:188173: Error: offset out of range
/tmp/camlasm480e0c.s:188177: Error: offset out of range
/tmp/camlasm480e0c.s:188194: Error: offset out of range
File "etc/uri_services_full.ml", line 1:
Error: Assembler error, input left in file /tmp/camlasm480e0c.s
```

**Comment by orbifx** [comment]
A possibly workaround for now: `opam install uri.1.9.2`

**Comment by dbuenzli** [comment]
I suspect this is not going to be fixed on this issue tracker. Something should be reported on mantis.

**Comment by orbifx** [comment]
Link?

**Comment by dbuenzli** [comment]
https://caml.inria.fr/mantis/

**Comment by orbifx** [comment]
Submitted: https://caml.inria.fr/mantis/view.php?id=7608

**Comment by DerRechner** [comment]
The Mantis tracker suggested that this might be fixed on 4.0.6+trunk. I just tested with Raspbian on a Raspberry Pi 3 and trunk. The installation of uri.1.3.12 worked.

**Comment by dbuenzli** [comment]
@gasche asked in the mantis issue if someone who sees the issue could test this with 4.06+dev.

**Comment by alrev** [comment]
I've tried to test uri 1.9.4 on Raspberry Pi 3 Model B (see message above) with 4.06.0+pr1250, but uri 1.9.4 depends on ppx_sexp_conv (>= v0.9.0), ppx_sexp_conv depends on ppx_core (>= v0.9 & < v0.10), ppx_core depends on ppx_ast (>= v0.9 & < v0.10), ppx_ast has ocaml-version < "4.06.0". Is there a way to overcome these dependencies?

**Comment by orbifx** [comment]
Any devs who can answer the dependencies question? I'm trying to build this too.

**Comment by orbifx** [comment]
Is there some way to by-pass the version checks in opam? I'm trying to build with `4.06.0+trunk`.

```
The following dependencies couldn't be met:
  - ppx_sexp_conv -> ppx_core < v0.10 -> ppx_ast < v0.10
  - ppx_sexp_conv -> ppx_tools >= 0.99.3
  - ppx_sexp_conv -> sexplib < 113.25.00
Your request can't be satisfied:
  - ppx_ast<v0.10 is not available because your system doesn't comply with ocaml-version >= "4.03.0" & ocaml-version < "4.06.0".
  - ppx_tools>=0.99.3 is not available because your system doesn't comply with ocaml-version >= "4.05" & ocaml-version < "4.06.0".
  - sexplib<113.25.00 is not available because your system doesn't comply with ocaml-version >= "4.02.3" & ocaml-version < "4.03.0".
```

**Comment by vitcozzolino** [comment]
I have the same issue on Cubietruck, with Ocaml version 4.04.2.

**Comment by alrev** [comment]
Tested today with 4.06.0+beta1+default-unsafe-string on Raspberry Pi 2 with Ubuntu MATE 16.04.3:
```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of uri failed at "jbuilder build -p uri -j 4".

#=== ERROR while installing uri.1.9.4 =========================================#
# opam-version 1.2.2
# os           linux
# command      jbuilder build -p uri -j 4
# path         /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4
# compiler     4.06.0+beta1+default-unsafe-string
# exit-code    1
# env-file     /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4/uri-3266-1872ec.env
# stdout-file  /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4/uri-3266-1872ec.out
# stderr-file  /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4/uri-3266-1872ec.err
### stderr ###
# Warning 58: no cmx file was found in path for module Toploop, and its interface was not compiled with -opaque
# [...]
#       ocamlc etc/uri_services_full.{cmo,cmt}
#       ocamlc etc/uri_services_full.cma
#     ocamlopt etc/uri_services_full.{cmx,o} (exit 2)
# (cd _build/default && /home/user1/.opam/4.06.0+beta1+default-unsafe-string/bin/ocamlopt.opt -w -40 -g -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/bytes -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/ocaml -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/re -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/sexplib -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/sexplib/0 -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/stringext -I lib -no-alias-deps -I etc -o etc/uri_services_full.cmx -c -impl etc/uri_services_full.ml)
# /tmp/camlasmb21d52.s: Assembler messages:
# /tmp/camlasmb21d52.s:52: Error: value of 68257 too large for field of 2 bytes at 30
# /tmp/camlasmb21d52.s:73429: Error: value of 68173 too large for field of 2 bytes at 204462
# File "etc/uri_services_full.ml", line 1:
# Error: Assembler error, input left in file /tmp/camlasmb21d52.s

```
Thanks to the new ppx_ast compatible with 4.06.0.
ppx_core and ppx_sexp_conv were --dev-repo.

**Comment by dbuenzli** [comment]
Thanks @alrev I reported your result on the bug upstream.

**Comment by gasche** [comment]
Thanks! (cc @xavierleroy: this seems to be an ARM issue that still exists in 4.06, reported in [MPR#7608](https://caml.inria.fr/mantis/view.php?id=7608)).

Could someone provide the following information for ease of reproducibility/testing?
- what does `ocamlc -config` say on a machine with this bug? (there are many ARM variants and compilation options)
- assuming someone that is not familiar with `--dev-repo` wants to reproduce, what's a sequence of command-line option to install the required `opam` dependencies? (Assuming one starts from a 4.06.0+... switch)

**Comment by alrev** [comment]
ocamlc -config:
```
version: 4.06.0+beta1
standard_library_default: /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/ocaml
standard_library: /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/ocaml
standard_runtime: /home/user1/.opam/4.06.0+beta1+default-unsafe-string/bin/ocamlrun
ccomp_type: cc
c_compiler: gcc
ocamlc_cflags: -O2 -fno-strict-aliasing -fwrapv -fPIC
ocamlopt_cflags: -O2 -fno-strict-aliasing -fwrapv
bytecomp_c_libraries: -lm  -ldl -lcurses -lpthread                  
native_c_libraries: -lm  -ldl
native_pack_linker: ld -r -o 
ranlib: ranlib
cc_profile: -pg
architecture: arm
model: default
int_size: 31
word_size: 32
system: linux_eabihf
asm: as
asm_cfi_supported: true
with_frame_pointers: false
ext_exe: 
ext_obj: .o
ext_asm: .s
ext_lib: .a
ext_dll: .so
os_type: Unix
default_executable_name: a.out
systhread_supported: true
host: armv7l-unknown-linux-gnueabihf
target: armv7l-unknown-linux-gnueabihf
profiling: true
flambda: false
spacetime: false
safe_string: false
default_safe_string: false
flat_float_array: true
windows_unicode: false
exec_magic_number: Caml1999X011
cmi_magic_number: Caml1999I021
cmo_magic_number: Caml1999O011
cma_magic_number: Caml1999A012
cmx_magic_number: Caml1999Y015
cmxa_magic_number: Caml1999Z014
ast_impl_magic_number: Caml1999M020
ast_intf_magic_number: Caml1999N018
cmxs_magic_number: Caml2007D002
cmt_magic_number: Caml2012T009
```

To get dev versions of ppx_core and ppx_sexp_conv:
```
opam pin add ppx_core --dev-repo
opam pin add ppx_sexp_conv --dev-repo
```
I used "default-unsafe-string" of compiler, because there are issues with some packages currently related to -safe-string option.

**Comment by gasche** [comment]
Thanks. While I'm at it: @alrev, supposing there is a bugfix one day, is there a name under which you would like to be credited for reporting help? (The standard is to use full names rather than pseudonyms, but it is as you are most comfortable.)

**Comment by xavierleroy** [comment]
The OCaml configuration is essentially the same as that of my two ARM 32 bit test platforms, which should make the detective work easier.  I'd be happy to help track down this issue but I'm scared by the huge set of dependencies.  (My OPAM kung-fu is weak.) Can someone come up with a self-contained repro case?  Or just attach the buggy `/tmp/camlasmXXXXXXX.s` file, that would already enable some post-mortem debugging.

**Comment by alrev** [comment]
camlasmbaf612.s, uri-19104-1872ec.err, uri-19104-1872ec.env:
 [camlasmbaf612.s_plus_err_and_env.zip](https://github.com/mirage/ocaml-uri/files/1381922/camlasmbaf612.s_plus_err_and_env.zip)
Log file:
```
=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of uri failed at "jbuilder build -p uri -j 4".

#=== ERROR while installing uri.1.9.4 =========================================#
# opam-version 1.2.2
# os           linux
# command      jbuilder build -p uri -j 4
# path         /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4
# compiler     4.06.0+beta1+default-unsafe-string
# exit-code    1
# env-file     /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4/uri-19104-1872ec.env
# stdout-file  /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4/uri-19104-1872ec.out
# stderr-file  /home/user1/.opam/4.06.0+beta1+default-unsafe-string/build/uri.1.9.4/uri-19104-1872ec.err
### stderr ###
# Warning 58: no cmx file was found in path for module Toploop, and its interface was not compiled with -opaque
# [...]
#       ocamlc etc/uri_services_full.{cmo,cmt}
#       ocamlc etc/uri_services_full.cma
#     ocamlopt etc/uri_services_full.{cmx,o} (exit 2)
# (cd _build/default && /home/user1/.opam/4.06.0+beta1+default-unsafe-string/bin/ocamlopt.opt -w -40 -g -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/bytes -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/ocaml -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/re -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/sexplib -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/sexplib/0 -I /home/user1/.opam/4.06.0+beta1+default-unsafe-string/lib/stringext -I lib -no-alias-deps -I etc -o etc/uri_services_full.cmx -c -impl etc/uri_services_full.ml)
# /tmp/camlasmbaf612.s: Assembler messages:
# /tmp/camlasmbaf612.s:52: Error: value of 68257 too large for field of 2 bytes at 30
# /tmp/camlasmbaf612.s:73429: Error: value of 68173 too large for field of 2 bytes at 204462
# File "etc/uri_services_full.ml", line 1:
# Error: Assembler error, input left in file /tmp/camlasmbaf612.s
```

**Comment by xavierleroy** [comment]
See the Mantis PR for an analysis and discussion of possible fixes.  The problem is exhibited by the autogenerated functions `Uri_services_full.udp_port_of_service` and `Uri_services_full.tcp_port_of_service` that are stupidly big and would benefit from being replaced by simple table lookups.

**Comment by gasche** [comment]
(I don't see anything on mantis [MPR#7608](https://caml.inria.fr/mantis/view.php?id=7608) yet, maybe the analysis/discussion comes later.)

Note that pattern-matching on many strings is very efficient since Benoît Vaugon's work on comparing word-by-word, and takes advantage of branch prediction. Using a hashtable may be noticeably slower in performance-sensitive applications (this may not apply to the Uri use-case, I don't know, but it can explain why people generate this sort of code) -- a small microbenchmark indicates that the lookup would be 6.5x slower on my machine if using `Hashtbl.find` instead.

**Comment by gasche** [comment]
Note that pattern-matching on many strings is very efficient since Benoît Vaugon's work on comparing word-by-word, and takes advantage of branch prediction. Using a hashtable may be noticeably slower in performance-sensitive applications (this may not apply to the Uri use-case, I don't know, but it can explain why people generate this sort of code) -- a small microbenchmark indicates that the lookup would be 6.5x slower on my machine if using Hashtbl.find instead.

**Comment by xavierleroy** [comment]
@gasche : micro-benchmarks don't take cache effects into account.  I estimate that the "stupidly big" functions I mentioned take 200 to 300 kbytes of ARM Thumb2 code each.  That's a lot less compact than the actual data.  Also, hashtables is the wrong implementation here: use binary search in a sorted array.

**Comment by xavierleroy** [comment]
A complete fix for this issue takes nontrivial work and is not going to happen for the forthcoming 4.06 release of OCaml.  

In the meantime you can get the code to compile on ARM by passing the `-fno-thumb` option to `ocamlopt`, either just on the problematic file, or on the whole library.  (The problem is specific to the Thumb instruction encoding.  Classic ARM instruction encoding should be fine.  ocamlopt defaults to Thumb encoding on ARMv7.  ARM-encoded and Thumb2-encoded codes interoperate just fine.)

I still suggest a different generated source code to avoid huge functions in the first place.

**Comment by avsm** [comment]
I might also just disable Uri_services_full from the default URI package, since it's rarely used to my knowledge and so not worth including in the base package. I can split it out into a separate uri-services package.

**Comment by gasche** [comment]
I played with the code generator a bit for the benchmarking above, and it wouldn't be very hard for me send a PR to change the code generation strategy. @avsm, would you merge that?

**Comment by oddodaoddo** [comment]
I can't install uri on an Intel i7 Mac either - see below.

Ognens-MacBook-Pro:~ ognend$ opam install uri
The following actions will be performed:
  ∗  install uri 1.9.4

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[uri] Archive in cache

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
[ERROR] The compilation of uri failed at "jbuilder build -p uri -j 4".

#=== ERROR while installing uri.1.9.4 =========================================#
# opam-version 1.2.2
# os           darwin
# command      jbuilder build -p uri -j 4
# path         /Users/ognend/.opam/system/build/uri.1.9.4
# compiler     system (4.06.0)
# exit-code    1
# env-file     /Users/ognend/.opam/system/build/uri.1.9.4/uri-27527-6d07ae.env
# stdout-file  /Users/ognend/.opam/system/build/uri.1.9.4/uri-27527-6d07ae.out
# stderr-file  /Users/ognend/.opam/system/build/uri.1.9.4/uri-27527-6d07ae.err
### stderr ###
# [...]
# and may change in future versions. (See manual section 8.5)
# File "lib/uri.ml", line 321, characters 27-44:
# Warning 52: Code should not depend on the actual values of
# this constructor's arguments. They are only for information
# and may change in future versions. (See manual section 8.5)
# File "lib/uri.ml", line 454, characters 11-12:
# Error: This expression has type bytes but an expression was expected of type
#          string
#
# Waiting for 1 job to finish.



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫
The following actions failed
  ∗  install uri 1.9.4
No changes have been performed

Result of running ocamlc -config:
Ognens-MacBook-Pro:~ ognend$ ocamlc -config
version: 4.06.0
standard_library_default: /usr/local/lib/ocaml
standard_library: /usr/local/lib/ocaml
standard_runtime: /usr/local/bin/ocamlrun
ccomp_type: cc
c_compiler: gcc
ocamlc_cflags: -O2 -fno-strict-aliasing -fwrapv
ocamlc_cppflags: -D_FILE_OFFSET_BITS=64 -D_REENTRANT
ocamlopt_cflags: -O2 -fno-strict-aliasing -fwrapv
ocamlopt_cppflags: -D_FILE_OFFSET_BITS=64 -D_REENTRANT
bytecomp_c_compiler: gcc -O2 -fno-strict-aliasing -fwrapv  -D_FILE_OFFSET_BITS=64 -D_REENTRANT
native_c_compiler: gcc -O2 -fno-strict-aliasing -fwrapv -D_FILE_OFFSET_BITS=64 -D_REENTRANT
bytecomp_c_libraries: -lcurses -lpthread
native_c_libraries:
native_pack_linker: ld -r -arch x86_64 -o
ranlib: ranlib
cc_profile: -pg
architecture: amd64
model: default
int_size: 63
word_size: 64
system: macosx
asm: clang -arch x86_64 -Wno-trigraphs -c
asm_cfi_supported: true
with_frame_pointers: false
ext_exe:
ext_obj: .o
ext_asm: .s
ext_lib: .a
ext_dll: .so
os_type: Unix
default_executable_name: a.out
systhread_supported: true
host: x86_64-apple-darwin16.7.0
target: x86_64-apple-darwin16.7.0
profiling: true
flambda: false
spacetime: false
safe_string: false
default_safe_string: true
flat_float_array: true
afl_instrument: false
windows_unicode: false
exec_magic_number: Caml1999X011
cmi_magic_number: Caml1999I022
cmo_magic_number: Caml1999O022
cma_magic_number: Caml1999A022
cmx_magic_number: Caml1999Y022
cmxa_magic_number: Caml1999Z022
ast_impl_magic_number: Caml1999M022
ast_intf_magic_number: Caml1999N022
cmxs_magic_number: Caml1999D022
cmt_magic_number: Caml1999T022

**Comment by gasche** [comment]
@oddodaoddo this is an unrelated failure. It can be fixed by a new ocaml-uri release, or by you using 4.05.0 instead. (As a temporary measure, using the switch `4.06.0+default-unsafe-string` should work and let you experiment with this package and 4.06, if you really need 4.06.)

**Comment by vbmithr** [comment]
I also need `ocaml-uri` on armv7.

@gasche : do you have a fix ready? I'd be happy to try.

**Comment by gasche** [comment]
I proposed a fix in #114, changing the code generation strategy to generate sorted array for bisection as @xavierleroy suggest. As a hint that this is a good idea, compilation time decreased from 45s to 4.5s on my machine.

(The generated code still contains large literal arrays, but I believe that those are ok. @vbmithr -- or anyone else -- I would appreciate if someone could to test on an ARM machine and report that the current code generation strategy is fine. If large arrays proved to be an issue, it would not be hard to change.)

**Comment by xavierleroy** [comment]
A temporary workaround is to compile with `ocamlopt -fno-thumb`, either the whole package or just the problematic files.

**Comment by alrev** [comment]
[#114](https://github.com/mirage/ocaml-uri/pull/114) can be compiled successfully with Raspberry Pi 3 and OCaml 4.04.0 (and the compilation is quite fast).

**Comment by avsm** [comment]
fixed in upstream ocaml and in uri>1.9.6 now.

**Comment by argent-smith** [comment]
❤️
